File: recursion_top_down.js
let arr = [1, 2, 3, 4, 5];

function sumRecursively(arr) {
  if (arr.length === 1) {
    return arr[0]
  } else {
    return arr[0] + sumRecursively(arr.slice(1));
  }
}

// console.log(sumRecursively(arr))

let string1 = "abcde"

function reverseStrRecursively(str) {
  if (str.length === 1) {
    return str[0]
  } else {
    return reverseStrRecursively(str.slice(1)) + str[0];
  }
}

// console.log(reverseStrRecursively(string1));

let string2 = "axbxcxd"

function countXsRecursively(str) {
  // if (str.length === 1) {
  //   if (str[0] === "x") {
  //     return 1
  //   } else {
  //     return 0
  //   }  
  // }
  if (str.length === 0) return 0;

  if (str[0] === "x") {
    return 1 + countXsRecursively(str.slice(1));
  } else {
    return countXsRecursively(str.slice(1));
  }
}

// console.log(countXsRecursively(string2));

function waysToClimb(numStairs) {
  if (numStairs === 1) return 1;
  if (numStairs === 2) return 2;
  if (numStairs === 3) return 4;
  if (numStairs === 4) return 7;
  return waysToClimb(numStairs - 1) + waysToClimb(numStairs - 2) + waysToClimb(numStairs - 3);
}

//console.log(waysToClimb(10));

function getAnagramsRecursively(str) {

}

let arr2 = ["ab", "c", "def", "ghij"];

function totalChars(arr) {
  if (arr.length === 1) {
    return arr[0].length;
  }
  return arr[0].length + totalChars(arr.slice(1));
}

//console.log(totalChars(arr2));

let arr3 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

function onlyEvens(arr) {
  if (arr.length === 1) {
    if (arr[0] % 2 === 0) {
      return [arr[0]];
    } else {
      return [];
    }    
  }

  if (arr[0] % 2 === 0) {
    let tempArr = onlyEvens(arr.slice(1));
    tempArr.unshift(arr[0]);
    return tempArr;
  } else {
    return onlyEvens(arr.slice(1));
  }
}

//console.log(onlyEvens(arr3))

function triangularNums(num) {
  if (num === 1) return 1;
  if (num === 2) return 3;
  if (num === 3) return 6;

  return num + triangularNums(num - 1);
}

//console.log(triangularNums(7));

function firstX(string, index = 0) {
  if (string[0] === "x") {
    return index;
  } else {
    return firstX(string.slice(1), index += 1);
  }
}

//console.log(firstX("abcdefghijklmnopqrstuvwxyz"))

function uniquePaths(rows, columns) {
  if (rows === 1 || columns === 1) {
    return 1;
  }
  return uniquePaths(rows - 1, columns) + uniquePaths(columns - 1, rows);
}

//console.log(uniquePaths(100, 2));


// can't get this...

function anagrams(str) {
 if (str.length === 1) return str[0];
 let collection = [];
 let anagramsOfSubstrings = anagrams(str.slice(1));
 
 for (let i = 0; i < anagramsOfSubstrings.length; i++) {
  let substr = anagramsOfSubstrings[i];

  for (let j = 0; j < substr.length; j++) {
    let newAnagram = substr.slice(0, j) + str[0] + substr.slice(j);
    collection.push(newAnagram);
  }
 }
 console.log(str);
 return collection;
}

console.log(anagrams('abcd'));

File: go_data_structures/recursion_top_down.go
package main

import "fmt"

func sumRecursively (s []int) (sum int) {
  if len(s) == 1 {
    return s[0]
  } else {
    return s[0] + sumRecursively(s[1:])
  }
}

func reverseStrRecursively (s string) (reversed string) {
  if len(s) == 1 {
    return string(s[0])
  } 
  return string(s[len(s) - 1:]) + reverseStrRecursively(s[:len(s) - 1])
}

func countXsRecursively (s string) (total int) {
  if len(s) == 1 {
    if string(s[0]) == "x" {
      return 1
    } else {
      return 0
    }
  }

  if string(s[0]) == "x" {
    return 1 + countXsRecursively(s[1:])
  } else {
    return countXsRecursively(s[1:])
  }
}

func waysToClimb(numStairs int) (int) {
  if numStairs == 0 {
    return 0
  }
  if numStairs == 1 {
    return 1
  }
  if numStairs == 2 {
    return 2
  }
  if numStairs == 3 {
    return 4
  }
  return waysToClimb(numStairs - 1) + waysToClimb(numStairs - 2) + waysToClimb(numStairs - 3)
}


func totalChars(s []string) int {
  if len(s) == 1 {
    return len(s[0])
  }
  return len(s[0]) + totalChars(s[1:])
}

func onlyEvens (n []int) (result []int) {
  if len(n) == 1 {
    if n[0] % 2 == 0 {
      result = []int{n[0]}
      return result
    } else {
      return onlyEvens(n[1:])
    }    
  }

  if n[0] % 2 == 0 {
    result = append(onlyEvens(n[1:]), n[0])
    return result
  } else {
    return onlyEvens(n[1:])
  }
}


func triangularNums(triNum int) int {
  if triNum == 1 {
    return 1
  }
  if triNum == 2 {
    return 3
  }
  if triNum == 3 {
    return 6
  }
  
  return triNum + triangularNums(triNum - 1)
}

func firstX(s string, currentIdx int) (idx int) {
  if string(s[0]) == "x" {
    return currentIdx
  } else {
    currentIdx += 1
    return firstX(s[1:], currentIdx)
  }
}

func uniquePaths(rows int, columns int) int {
  if rows == 1 || columns == 1 {
    return 1
  }
  return uniquePaths(rows - 1, columns) + uniquePaths(rows, columns - 1)
}

func main() {

  // toSum := []int{1, 2, 3, 4, 5}
  // fmt.Println(sumRecursively(toSum))

  //fmt.Println(reverseStrRecursively("abcde"))
  //fmt.Println(countXsRecursively("axbxcxd"))
  //fmt.Println(waysToClimb(10))
  //fmt.Println(uniquePaths(10, 10))
  //fmt.Println(totalChars([]string{"ab", "c", "def", "ghij"}))
  // fmt.Println(onlyEvens([]int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}))
  // fmt.Println(triangularNums(7))
  fmt.Println(firstX("abcdefghijklmnopqrstuvwxyz", 0))

}